### DISCUSSION TITLE
Version 2.1.2 (incl. Python bindings)

### DISCUSSION THREAD

**Discussion by saran-t** [google-deepmind/mujoco#185] [Announcements]
## New modules
1. Added new [Python bindings](https://mujoco.readthedocs.io/en/latest/python.html), which can be installed via `pip install mujoco`, and imported as `import mujoco`.
2. Added new [Unity plug-in](https://mujoco.readthedocs.io/en/latest/unity.html).
3. Added a new `introspect` module, which provides reflection-like capability for MuJoCo's public API, currently describing functions and enums. While implemented in Python, this module is expected to be generally useful for automatic code generation targeting multiple languages. (This is not shipped as part of the `mujoco` Python bindings package.)

## API changes
4. Moved definition of `mjtNum` floating point type into a new header [mjtnum.h](https://github.com/deepmind/mujoco/blob/2.1.2/include/mjtnum.h).
5. Renamed header `mujoco_export.h` to [mjexport.h](https://github.com/deepmind/mujoco/blob/2.1.2/include/mjexport.h).
6. Added `mj_printFormattedData`, which accepts a format string for floating point numbers, for example to increase precision.

## General
7. MuJoCo can load [OBJ](https://en.wikipedia.org/wiki/Wavefront_.obj_file) mesh files.
   - Meshes containing polygons with more than 4 vertices are not supported.
   - In OBJ files containing multiple object groups, any groups after the first one will be ignored.
8. Added optional frame-of-reference specification to: [`framepos`](https://mujoco.readthedocs.io/en/latest/XMLreference.html#sensor-framepos), [`framequat`](https://mujoco.readthedocs.io/en/latest/XMLreference.html#sensor-framequat), [`framexaxis`](https://mujoco.readthedocs.io/en/latest/XMLreference.html#sensor-framexaxis), [`frameyaxis`](https://mujoco.readthedocs.io/en/latest/XMLreference.html#sensor-frameyaxis), [`framezaxis`](https://mujoco.readthedocs.io/en/latest/XMLreference.html#sensor-framezaxis), [`framelinvel`](https://mujoco.readthedocs.io/en/latest/XMLreference.html#sensor-framelinvel), and [`frameangvel`](https://mujoco.readthedocs.io/en/latest/XMLreference.html#sensor-frameangvel) sensors. The frame-of-reference is specified by new `reftype` and `refname` attributes.
9. Sizes of [user parameters](https://mujoco.readthedocs.io/en/latest/modeling.html#cuser) are now automatically inferred.
   - Declarations of user parameters in the top-level [size](https://mujoco.readthedocs.io/en/latest/XMLreference.html#size) clause (e.g. `nuser_body`, `nuser_jnt`, etc.) now accept a value of -1, which is the default. This will automatically set the value to the length of the maximum associated `user` attribute defined in the model.
   - Setting a value smaller than -1 will lead to a compiler error (previously a segfault).
   - Setting a value to a length smaller than some `user` attribute defined in the model will lead to an error (previously additional values were ignored).
10. Increased the maximum number of lights in an [`mjvScene`](https://mujoco.readthedocs.io/en/latest/APIreference.html#mjvscene) from 8 to 100.
11. Saved XML files only contain explicit [inertial](https://mujoco.readthedocs.io/en/latest/XMLreference.html#inertial) elements if the original XML included them. Inertias that were automatically inferred by the compiler's [inertiafromgeom](https://mujoco.readthedocs.io/en/latest/XMLreference.html#compiler) mechanism remain unspecified.
12. User-selected geoms are always rendered as opaque. This is useful in interactive visualizers.
13. Static geoms now respect their [geom group](https://mujoco.readthedocs.io/en/latest/XMLreference.html#geom) for visualisation. Until this change rendering of static geoms could only be toggled using the [`mjVIS_STATIC`](https://mujoco.readthedocs.io/en/latest/APIreference.html#mjtvisflag) visualisation flag . After this change, both the geom group and the visualisation flag need to be enabled for the geom to be rendered.
14. Pointer parameters in function declarations in [mujoco.h](https://github.com/deepmind/mujoco/blob/2.1.2/include/mujoco.h) that are supposed to represent fixed-length arrays are now spelled as arrays with extents, e.g. `mjtNum quat[4]` rather than `mjtNum* quat`. From the perspective of C and C++, this is a non-change since array types in function signatures decay to pointer types. However, it allows autogenerated code to be aware of expected input shapes.
15. Experimental stateless fluid interaction model. As described [here](https://mujoco.readthedocs.io/en/latest/computation.html?highlight=passive#passive-forces), fluid forces use sizes computed from body inertia. While sometimes convenient, this is very rarely a good approximation. In the new model forces act on geoms, rather than bodies, and have a several user-settable parameters. The model is activated by setting a new attribute: `<geom fluidshape="ellipsoid"/>`. The parameters are described succinctly [here](https://mujoco.readthedocs.io/en/latest/XMLreference.html?highlight=geom#body-geom), but we leave a full description or the model and its parameters to when this feature leaves experimental status.

## Bug Fixes
16. `mj_loadXML` and `mj_saveLastXML` are now locale-independent. The Unity plugin should now work correctly for users whose system locales use commas as decimal separators.
17.  XML assets in VFS no longer need to end in a null character. Instead, the file size is determined by the size parameter of the corresponding VFS entry.
18.  Fix a vertex buffer object memory leak in `mjrContext` when skins are used.
19.  Camera quaternions are now normalized during XML compilation.

## Binary build
20. Windows binaries are now built with Clang.

<hr /><em>This discussion was created from the release <a href='https://github.com/deepmind/mujoco/releases/tag/2.1.2'>2.1.2</a>.</em>

**Comment by rohanpsingh** [comment]
Thanks a lot for releasing the python bindings 🎉  
Looking forward to the open-source release!

I’ve been using mujoco-py but now want to migrate to mujoco.

However, I’ve been unable to render the scene, even after following the instructions [here](https://mujoco.readthedocs.io/en/latest/python.html#rendering).
Do you have a minimal example, that shows how to render?

**Reply by saran-t** [reply]
Here you go: https://colab.research.google.com/drive/1d7r7_TFXLBsnUfsYz77vye4jKhSEXA0t?usp=sharing

~There's some weirdness in OpenGL library loading, but so far I've only found this to be a problem when running on Colab.~
EDIT: The copy of `libglewegl.so` that's bundled with `2.1.2.post1` Linux wheels has a `DT_NEEDED` entry on `libOpenGL.so.0`, so the OpenGL loading issue that was present earlier today is now resolved.

**Comment by BolunDai0216** [comment]
Is there a plan to write more detailed documentations for the python bindings?

**Reply by saran-t** [reply]
Do you have any specific aspect that you'd particularly like to see in the documentation?

**Reply by BolunDai0216** [reply]
A one-to-one correspondence with what is mentioned in the C API reference would be awesome. Some example code for running a control loop would be also be great.

**Reply by saran-t** [reply]
We're unlikely to rewrite the entire documentation to specifically target Python. The Python section in our docs is meant to serve as a sort of translation guide for how you can convert various aspects of the C API into Python, but the C documentation remains authoritative.

Having said that, we'll consider expanding our minimal example to also include setting control signals.

**Reply by BolunDai0216** [reply]
Would it also be possible to add an example showing how rendering can be done in Python?

**Reply by saran-t** [reply]
We might add the example in my Colab notebook in the comment above into our docs. Thanks for the feedback!

**Comment by pab47** [comment]
I tried (unsuccessfully) running the aarch64 v 2.1.2 on raspberry pi. I have libglfw3-dev installed. 

On running make in the sample folder, I get /usr/bin/ld:../lib/libmujoco_nogl.so: file format not recognized; treating as a linker script; and /usr/bin/ld:../lib/libmujoco_nogl.s0:2:syntax error. 

I tried to search for similar errors on the web with no luck. Any suggestions?
